### PR TITLE
[java] New rule PrimitiveWrapperInstantiation

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

This new rule merges the rules
* ByteInstantiation
* IntegerInstantiation
* LongInstantiation
* ShortInstantiation
* BooleanInstantiation (this does a bit more - encourages to use boolean constants instead of new boolean or valueof)

The rule may have some overlap with #3364, if the constructor call is unnecessary because of autoboxing or some other case caught by #3364. This looks ok though 

This also deprecates the rule [UnnecessaryWrapperObjectCreation](https://pmd.github.io/latest/pmd_rules_java_performance.html#unnecessarywrapperobjectcreation) because it's useless, as described in #3364. The replacement hinted at in the updated rule description is #3364.

## Related issues

- Relates to #2973. Together with #3364 they fix that issue.

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset changes 0 violations,<br />
introduces 484 new violations, 0 new errors and 0 new configuration errors,<br />
removes 395 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/c815159746ac4cd2b9a34b147d13e605/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/PrimitiveWrapperInstantiation.xml [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    public void go(){
+        foo(new Boolean("true"));
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>don't use new Boolean() in method call</description>
+        <expected-problems>0</expected-problems>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/PrimitiveWrapperInstantiation.xml:83]: regressionTest="false" and expected-problems=1

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/PrimitiveWrapperInstantiation.xml [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    static {
+        foo(Boolean.TRUE);
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>don't use new Boolean() in static block</description>
+        <expected-problems>0</expected-problems>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/PrimitiveWrapperInstantiation.xml:119]: regressionTest="false" and expected-problems=1

**Review by adangel — APPROVED** [review]
Looks good, thanks!

I know now, what the problem will be, when we integrate BooleanInstantiation into this rule: for Boolean, we actually need a very different violation message: We don't want to say "Do not use new Boolean(...), prefer Boolean.valueOf(...)", instead we want e.g. "Do not use new Boolean(...) or Boolean.valueOf(...), prefer Boolean.TRUE or Boolean.FALSE".
Keeping it as a XPath rule, this might not be possible. But if we reimplement this as a simple Java rule, we can issue different messages.

Maybe I'll have a look at this on Thursday - otherwise we can integrate BooleanInstantiation later.

*Reminder:* We need to mention the deprecated rules in the release notes - and of course also the new rule.

**Pull request merged**